File: finance-tracker-backend/src/main/java/com/finance/tracker/accounts/domain/AccountType.java
package com.finance.tracker.accounts.domain;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import lombok.Getter;

@Getter
public enum AccountType {

    // Assets
    BANK(true),
    SAVINGS(true),
    CHECKING(true),
    CASH(true),
    WALLET(true),
    INVESTMENT(true),
    UNKNOWN(true),

    // Liabilities
    CREDIT_CARD(false),
    LOAN(false);

    private final boolean isAsset;

    AccountType(boolean isAsset) {
        this.isAsset = isAsset;
    }

    public boolean isAssetType() {
        return isAsset;
    }

    public boolean isLiabilityType() {
        return !isAsset;
    }

    @JsonValue
    public String toJson() {
        return name();
    }

    @JsonCreator
    public static AccountType fromString(String value) {
        if (value == null || value.isBlank()) {
            return UNKNOWN;
        }
        String normalized = value.trim().toUpperCase().replace(" ", "_").replace("-", "_");
        for (AccountType type : values()) {
            if (type.name().equalsIgnoreCase(normalized)) {
                return type;
            }
        }
        return UNKNOWN;
    }
}

File: finance_app/lib/features/account/data/model/account_type.dart
enum AccountType {
  bank,
  savings,
  checking,
  cash,
  creditCard,
  loan,
  investment,
  unknown,
}

extension AccountTypeX on AccountType {
  String get label {
    switch (this) {
      case AccountType.bank:
        return "Bank Account";
      case AccountType.savings:
        return "Savings Account";
      case AccountType.checking:
        return "Checking Account";
      case AccountType.cash:
        return "Cash";
      case AccountType.creditCard:
        return "Credit Card";
      case AccountType.loan:
        return "Loan";
      case AccountType.investment:
        return "Investment";
      case AccountType.unknown:
        return "Unknown";
    }
  }

  String get apiValue {
    switch (this) {
      case AccountType.bank:
        return "BANK";
      case AccountType.savings:
        return "SAVINGS";
      case AccountType.checking:
        return "CHECKING";
      case AccountType.cash:
        return "CASH";
      case AccountType.creditCard:
        return "CREDIT_CARD";
      case AccountType.loan:
        return "LOAN";
      case AccountType.investment:
        return "INVESTMENT";
      case AccountType.unknown:
        return "UNKNOWN";
    }
  }
}

extension AccountTypeParser on String {
  AccountType toAccountType() {
    final normalized = trim().toUpperCase().replaceAll(" ", "_");

    switch (normalized) {
      case "BANK":
        return AccountType.bank;
      case "SAVINGS":
        return AccountType.savings;
      case "CHECKING":
        return AccountType.checking;
      case "CASH":
        return AccountType.cash;
      case "CREDIT_CARD":
        return AccountType.creditCard;
      case "LOAN":
        return AccountType.loan;
      case "INVESTMENT":
        return AccountType.investment;
      default:
        return AccountType.unknown;
    }
  }
}
